controllers/events/trigger_controller.go

```diff
@@ -0,0 +1,392 @@
+		return ctrl.Result{}, util.IgnoreNotFound(err)
+	}
+
+	if _, err := r.createOrUpdateTrigger(&trigger); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// createOrUpdateTrigger will do:
+// 1. Generate a dapr component specification for the EventBus associated with the Trigger
+//    and create the dapr component (will check if it needs to be updated)
+//    and set the TriggerConfig.EventBusComponentName, TriggerConfig.EventBusTopic, TriggerConfig.EventBusSpecEncode.
+// 2. Generate dapr component specifications for the subscribers
+//    and create the dapr components (will check if they need to be updated)
+//    and set the TriggerConfig.Subscribers, TriggerConfig.SinkSpecEncode.
+// 3. Create a workload for Trigger (will check if they need to be updated)
+//    and pass in the TriggerConfig as an environment variable.
+func (r *TriggerReconciler) createOrUpdateTrigger(trigger *openfunctionevent.Trigger) (ctrl.Result, error) {
+	_ = r.Log.WithName("createOrUpdate")
+
+	// Handle EventBus reconcile.
+	if trigger.Spec.EventBus != nil {
+		if err := r.handleEventBus(trigger); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Handle Subscriber reconcile.
+	if err := r.handleSubscriber(trigger); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *TriggerReconciler) handleEventBus(trigger *openfunctionevent.Trigger) error {
+
+	// Retrieve the specification of EventBus associated with the EventSource
+	eventBus := retrieveEventBus(r.ctx, r.Client, trigger.Namespace, trigger.Spec.EventBus.Name)
+	if eventBus == nil {
+		return errors.New("cannot retrieve eventBus")
+	}
+
+	componentName := fmt.Sprintf(TriggerBusComponentNameTmpl, trigger.Name)
+
+	// Set SourceConfig.EventBusComponentName and SourceConfig.EventBusTopic.
+	r.envs.EventBusComponentName = componentName
+	r.envs.EventBusTopic = trigger.Spec.EventBus.Topic
+
+	// Generate a dapr component based on the specification of EventBus.
+	component := &componentsv1alpha1.Component{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      componentName,
+			Namespace: trigger.Namespace,
+		},
+	}
+
+	if eventBus.Spec.Nats != nil {
+		specBytes, err := json.Marshal(eventBus.Spec.Nats)
+		if err != nil {
+			return err
+		}
+
+		// Set the TriggerConfig.EventBusSpecEncode which will reflect the specification content changes of dapr component back to the Trigger workload,
+		// informing it that it needs to rebuild.
+		r.envs.EventBusSpecEncode = base64.StdEncoding.EncodeToString(specBytes)
+
+		// Create the dapr component for Trigger to retrieve event from EventBus.
+		if _, err := ctrl.CreateOrUpdate(r.ctx, r.Client, component, mutateDaprComponent(r.Scheme, component, trigger, eventBus.Spec.Nats)); err != nil {
+			return err
+		}
+	} else {
+		return nil
+	}
+	return nil
+}
+
+func (r *TriggerReconciler) handleSubscriber(trigger *openfunctionevent.Trigger) error {
+	var components []*componentsv1alpha1.Component
```

Comment:
use map instead of slice

Reply:
Has been optimized to use the `map` instead of `slice`